### PR TITLE
Create MS.SqlTools.Credentials project

### PR DESCRIPTION
**Pull request by kburtram**
Create an initial project for the credentials service provider executable.

### CODE REVIEW

File: src/Microsoft.SqlTools.Credentials/Program.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+    /// <summary>
+    /// Main application class for Credentials Service Host executable
+    /// </summary>
+    internal class Program
+    {
+        /// <summary>
+        /// Main entry point into the Credentials Service Host
+        /// </summary>
+        internal static void Main(string[] args)
```
Comment (benrr101) [inline src/Microsoft.SqlTools.Credentials/Program.cs]: Is the goal of this to create a new service host that will implement its own JSON RPC just for servicing credentials requests? Or is the goal to spin the credentials service out to its own library that the tools service (and potentially other tools services) would consume? I think if we can go the latter route we'll reduce overhead tremendously and drastically reduce the amount of duplicated code.

File: src/Microsoft.SqlTools.Credentials/Program.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            CommandOptions commandOptions = new CommandOptions(args);
+            if (commandOptions.ShouldExit)
+            {
+                return;
+            }
+
+            // turn on Verbose logging during early development
+            // we need to switch to Normal when preparing for public preview
+            Logger.Initialize(minimumLogLevel: LogLevel.Verbose, isEnabled: commandOptions.EnableLogging);
+            Logger.Write(LogLevel.Normal, "Starting SQL Tools Service Host");
```
Comment (benrr101) [inline src/Microsoft.SqlTools.Credentials/Program.cs]: SqlTools Credentials Provider

File: src/Microsoft.SqlTools.Credentials/Program.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.SqlTools.Credentials.Hosting;
+using Microsoft.SqlTools.Credentials.Utility;
+using Microsoft.SqlTools.ServiceLayer.SqlContext;
+using Microsoft.SqlTools.ServiceLayer.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer
```
Comment (benrr101) [inline src/Microsoft.SqlTools.Credentials/Program.cs]: Microsoft.SqlTools.Credentials

File: src/Microsoft.SqlTools.Credentials/project.json [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        "../Microsoft.SqlTools.ServiceLayer/Utility/ThreadSynchronizationContext.cs",
+        "../Microsoft.SqlTools.ServiceLayer/Utility/Validate.cs",
+      ]
+    }
+  },  
+  "dependencies": {
+    "Newtonsoft.Json": "9.0.1",
+    "System.Data.Common": "4.1.0",
+    "System.Data.SqlClient": "4.4.0-sqltools-24613-04",
+    "Microsoft.SqlServer.Smo": "140.1.12",
```
Comment (benrr101) [inline src/Microsoft.SqlTools.Credentials/project.json]: I'm pretty sure we can clean up these dependencies.

File: src/Microsoft.SqlTools.Credentials/project.json [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  "version": "1.0.0-*",
+  "buildOptions": {
+    "debugType": "portable",
+    "emitEntryPoint": true,
+    "embed": [
+      "../Microsoft.SqlTools.ServiceLayer/Localization/sr.es.resx",
+      "../Microsoft.SqlTools.ServiceLayer/Localization/sr.resx"
+    ],    
+    "compile": {
+      "exclude": "*.*",
```
Comment (benrr101) [inline src/Microsoft.SqlTools.Credentials/project.json]: If we're going the route of having this be its own standalone JSON RPC service, we need to make certain that we decouple it from the ServiceLayer project. I strongly encourage you to factor out the /Hosting/ folder into its own library that is shared between the two services. When I did the major refactor back in July of the code that came from the PowerShell dude, I had this exact scenario in mind.

Comment (benrr101) [inline src/Microsoft.SqlTools.Credentials/project.json]: Generally speaking, I'm pretty sure project.json files are meant to be a blacklist versus a whitelist of what files to compile.

File: src/Microsoft.SqlTools.Credentials/Program.cs (hunk 1 above)
Comment (kburtram) [inline src/Microsoft.SqlTools.Credentials/Program.cs]: the goal is to make this its own service application to handle credentials.

File: src/Microsoft.SqlTools.Credentials/Microsoft.SqlTools.Credentials.xproj [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+  </PropertyGroup>
+  <Import Project="$(VSToolsPath)\DotNet\Microsoft.DotNet.Props" Condition="'$(VSToolsPath)' != ''" />
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>{0D61DC2B-DA66-441D-B9D0-F76C98F780F9}</ProjectGuid>
+    <RootNamespace>Microsoft.SqlTools.ServiceLayer</RootNamespace>
```
Comment (kburtram) [inline src/Microsoft.SqlTools.Credentials/Microsoft.SqlTools.Credentials.xproj]: >SqlTools.ServiceLayer [](start = 29, length = 21)

need to update this.

**Comment by kburtram** [comment]
Code coverage drop is related to bringing so much new code for localization and service start-up.  I'll address the drop as a separate PR.

**Comment by kburtram** [comment]
This is a large refactoring which mostly moves already reviewed code around.  I'll push it up now and then look into coveralls separately.  I've you have feedback on the change please leave them on the commit and email me.

**Pull request merged**